cs/src/core/Device/ManagedLocalStorageDevice.cs

```diff
@@ -97,65 +104,107 @@ private void RecoverFiles()
-#if NETSTANDARD
-                    if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    // FileStream.WriteAsync is not thread-safe hence need a lock here
+                    lock (this)
+                    {
+                        readTask = logReadHandle.ReadAsync(umm.Memory).AsTask();
+#else
+                    memory = pool.Get((int)readLength);
+                    lock (this)
+                    {
+                        readTask = logReadHandle.ReadAsync(memory.buffer, 0, (int)readLength);
+#endif
+                    }
+                    if (IsWindows)
                     {
-                        if (offset >= 0) streampool?.Return(offset);
+                        // If non-netstandard (==windows-only), or netstandard+windows, we can return to the pool immediately.
+                        // For netstandard+linux we will return after the operation completes.
+                        streampool?.Return(offset);
                     }
+                }
+                catch
+                {
+                    Interlocked.Decrement(ref numPending);
+#if !NETSTANDARD2_1
+                    memory?.Return();
 #endif
+                    streampool?.Return(offset);
+                    streampool?.DisposeIfNeeded(offset, logReadHandle);
+                    callback(uint.MaxValue, 0, context);
+                    return;
+                }
+                
+                try
+                {
+                    numBytes = await readTask;
 
-                    callback(errorCode, (uint)t.Result, context);
+                    // Sequentialize all reads from same handle on non-windows
+                    if (!IsWindows)
+                    {
+                        streampool?.Return(offset);
+                    }
+                    streampool?.DisposeIfNeeded(offset, logReadHandle);
                 }
-                );
+                catch (Exception ex)
+                {
+                    if (ex.InnerException != null &&
+                        ex.InnerException is IOException ioex)
+                    {
+                        errorCode = (uint)(ioex.HResult & 0x0000FFFF);
+                    }
+                    else
+                    {
+                        errorCode = uint.MaxValue;
+                    }
 
-#if NETSTANDARD
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-                if (offset >= 0) streampool?.Return(offset);
-#else
-            if (offset >= 0) streampool?.Return(offset);
+                    numBytes = 0;
+                }
+                finally
+                {
+                    Interlocked.Decrement(ref numPending);
+#if !NETSTANDARD2_1
+                    unsafe
+                    {
+                        fixed (void* source = memory.buffer)
+                            Buffer.MemoryCopy(source, (void*)destinationAddress, numBytes, numBytes);
```

Comment:
Why should this memory copy happen when 'await readTask' throws an exception?

Reply:
Fixed, please re-check code.